lib/jsonapi/deserializable/resource.rb

```diff
@@ -53,53 +75,54 @@ def deserialize!
       end
 
       def deserialize_type!
-        return unless @type && self.class.type_block
         instance_exec(@type, &self.class.type_block)
       end
 
       def deserialize_id!
-        return unless @id && self.class.id_block
+        return unless @id
         instance_exec(@id, &self.class.id_block)
       end
 
       def deserialize_attrs!
-        self.class.attr_blocks.each do |attr, block|
-          next unless @attributes.key?(attr)
-          instance_exec(@attributes[attr], &block)
+        @attributes.each do |key, val|
+          if self.class.attr_blocks.key?(key)
+            instance_exec(val, &self.class.attr_blocks[key])
+          else
+            instance_exec(key, val, &self.class.default_attr_block)
+          end
         end
       end
 
       def deserialize_rels!
-        deserialize_has_one_rels!
-        deserialize_has_many_rels!
-      end
-
-      def deserialize_has_one_rels!
-        self.class.has_one_rel_blocks.each do |key, block|
-          rel = @relationships[key]
-          next unless rel && (rel['data'].nil? || rel['data'].is_a?(Hash))
-          deserialize_has_one_rel!(rel, &block)
+        @relationships.each do |key, val|
+          if val['data'].is_a?(Array)
+            deserialize_has_many_rel!(key, val)
+          else
+            deserialize_has_one_rel!(key, val)
+          end
         end
       end
 
-      def deserialize_has_one_rel!(rel, &block)
-        id = rel['data'] && rel['data']['id']
-        type = rel['data'] && rel['data']['type']
-        instance_exec(rel, id, type, &block)
-      end
-
-      def deserialize_has_many_rels!
-        self.class.has_many_rel_blocks.each do |key, block|
-          rel = @relationships[key]
-          next unless rel && rel['data'].is_a?(Array)
-          deserialize_has_many_rel!(rel, &block)
+      def deserialize_has_one_rel!(key, val)
+        id   = val['data'] && val['data']['id']
+        type = val['data'] && val['data']['type']
+        if self.class.has_one_rel_blocks.key?(key)
+          instance_exec(val, id, type, &self.class.has_one_rel_blocks[key])
```

Comment:
`#field` makes sense. I guess I have no problems, but personally I usually do something like

```ruby
dsl = MyDSL.new(self)
dsl.instance_eval(&blk)
```

That way this object can separate its actual responsibilities (read configuration and perform work based on input) versus the DSL responsibilities (generate configuration/metadata, throw errors, etc). YMMV, just throwing it out there.